File: gitdot-web/packages/api/src/resource/repository.ts
import { z } from "zod";

import { BuildResource } from "./build";
import { QuestionResource } from "./question";
import { ReviewResource } from "./review";
import { CommitFilterResource } from "./settings";

export const SyntaxHighlight = z.enum([
  "delimiter",
  "normal",
  "string",
  "type",
  "comment",
  "keyword",
  "tree_sitter_error",
]);
export type SyntaxHighlight = z.infer<typeof SyntaxHighlight>;

export const DiffChangeResource = z.object({
  start: z.number().int(),
  end: z.number().int(),
  content: z.string(),
  highlight: SyntaxHighlight,
});
export type DiffChangeResource = z.infer<typeof DiffChangeResource>;

export const DiffLineResource = z.object({
  line_number: z.number().int(),
  changes: z.array(DiffChangeResource),
});
export type DiffLineResource = z.infer<typeof DiffLineResource>;

export const DiffPairResource = z.object({
  lhs: DiffLineResource.optional(),
  rhs: DiffLineResource.optional(),
});
export type DiffPairResource = z.infer<typeof DiffPairResource>;

export const DiffHunkResource = z.array(DiffPairResource);
export type DiffHunkResource = z.infer<typeof DiffHunkResource>;

export const RepositoryFileResource = z.object({
  commit_sha: z.string(),
  path: z.string(),
  sha: z.string(),
  content: z.string(),
  encoding: z.string(),
});
export type RepositoryFileResource = z.infer<typeof RepositoryFileResource>;

export const CommitAuthorResource = z.object({
  id: z.uuid().optional(),
  name: z.string(),
  email: z.string(),
});
export type CommitAuthorResource = z.infer<typeof CommitAuthorResource>;

export const RepositoryDiffStatResource = z.object({
  path: z.string(),
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
});
export type RepositoryDiffStatResource = z.infer<
  typeof RepositoryDiffStatResource
>;

export const RepositoryDiffFileResource = z.object({
  path: z.string(),
  lines_added: z.number().int(),
  lines_removed: z.number().int(),
  hunks: z.array(DiffHunkResource),
  left_content: z.string().optional(),
  right_content: z.string().optional(),
});
export type RepositoryDiffFileResource = z.infer<
  typeof RepositoryDiffFileResource
>;

export const RepositoryCommitResource = z.object({
  owner_name: z.string(),
  repo_name: z.string(),
  sha: z.string(),
  parent_sha: z.string(),
  message: z.string(),
  date: z.iso.datetime(),
  author: CommitAuthorResource,
  review_number: z.number().int().optional(),
  diff_position: z.number().int().optional(),
  diffs: z.array(RepositoryDiffStatResource),
});
export type RepositoryCommitResource = z.infer<typeof RepositoryCommitResource>;

export const RepositoryCommitsResource = z.object({
  commits: z.array(RepositoryCommitResource),
});
export type RepositoryCommitsResource = z.infer<
  typeof RepositoryCommitsResource
>;

export const RepositoryCommitDiffResource = z.object({
  sha: z.string(),
  parent_sha: z.string(),
  files: z.array(RepositoryDiffFileResource),
});
export type RepositoryCommitDiffResource = z.infer<
  typeof RepositoryCommitDiffResource
>;

export const PathType = z.enum(["blob", "tree", "commit", "unknown"]);
export type PathType = z.infer<typeof PathType>;

export const RepositoryPathResource = z.object({
  path: z.string(),
  name: z.string(),
  path_type: PathType,
  sha: z.string(),
});
export type RepositoryPathResource = z.infer<typeof RepositoryPathResource>;

export const RepositoryPathsResource = z.object({
  ref_name: z.string(),
  commit_sha: z.string(),
  entries: z.array(RepositoryPathResource),
});
export type RepositoryPathsResource = z.infer<typeof RepositoryPathsResource>;

export const RepositoryFolderResource = z.object({
  type: z.literal("folder"),
  commit_sha: z.string(),
  path: z.string(),
  entries: z.array(RepositoryPathResource),
});
export type RepositoryFolderResource = z.infer<typeof RepositoryFolderResource>;

export const RepositoryBlobResource = z.discriminatedUnion("type", [
  RepositoryFileResource.extend({ type: z.literal("file") }),
  RepositoryFolderResource,
]);
export type RepositoryBlobResource = z.infer<typeof RepositoryBlobResource>;

export const RepositoryBlobsResource = z.object({
  blobs: z.array(RepositoryBlobResource),
});
export type RepositoryBlobsResource = z.infer<typeof RepositoryBlobsResource>;

export const RepositoryBlobDiffsResource = z.object({
  diffs: z.record(z.string(), RepositoryDiffFileResource),
});
export type RepositoryBlobDiffsResource = z.infer<
  typeof RepositoryBlobDiffsResource
>;

export const RepositoryResource = z.object({
  id: z.uuid(),
  name: z.string(),
  owner: z.string(),
  visibility: z.string(),
  description: z.string().optional(),
  stars: z.number().int().nonnegative(),
  readonly: z.boolean(),
  created_at: z.iso.datetime(),
});
export type RepositoryResource = z.infer<typeof RepositoryResource>;

export const RepositorySettingsResource = z.object({
  commit_filters: z.array(CommitFilterResource).optional(),
});
export type RepositorySettingsResource = z.infer<
  typeof RepositorySettingsResource
>;

export const RepositoryQuestionsResource = z.object({
  questions: z.array(QuestionResource),
});
export type RepositoryQuestionsResource = z.infer<
  typeof RepositoryQuestionsResource
>;

export const RepositoryReviewsResource = z.object({
  reviews: z.array(ReviewResource),
});
export type RepositoryReviewsResource = z.infer<
  typeof RepositoryReviewsResource
>;

export const RepositoryBuildsResource = z.object({
  builds: z.array(BuildResource),
});
export type RepositoryBuildsResource = z.infer<typeof RepositoryBuildsResource>;

export const RepositoryResourcesResource = z.object({
  last_commit: z.string(),
  last_updated: z.iso.datetime().optional(),
  paths: RepositoryPathsResource.optional(),
  commits: RepositoryCommitsResource.optional(),
  blobs: RepositoryBlobsResource.optional(),
  questions: RepositoryQuestionsResource.optional(),
  reviews: RepositoryReviewsResource.optional(),
  builds: RepositoryBuildsResource.optional(),
});
export type RepositoryResourcesResource = z.infer<
  typeof RepositoryResourcesResource
>;

File: gitdot-server/src/dto/repository.rs
use gitdot_api::resource::repository as api;
use gitdot_core::{
    dto::{
        CommitAuthorResponse, CommitResponse, PathType, RepositoryActivityEvent,
        RepositoryBlobDiffsResponse, RepositoryBlobPairResponse, RepositoryBlobResponse,
        RepositoryBlobsResponse, RepositoryCommitFilterResponse, RepositoryCommitResponse,
        RepositoryCommitsResponse, RepositoryDiffFileResponse, RepositoryFileResponse,
        RepositoryFolderResponse, RepositoryPath, RepositoryPathsResponse, RepositoryResponse,
    },
    model::CommitDiff,
};

use super::IntoApi;

impl IntoApi for RepositoryResponse {
    type ApiType = api::RepositoryResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryResource {
            id: self.id,
            name: self.name,
            owner: self.owner,
            visibility: self.visibility,
            description: self.description,
            stars: self.stars,
            user_star: self.user_star,
            readonly: self.readonly,
            created_at: self.created_at,
        }
    }
}

impl IntoApi for RepositoryCommitsResponse {
    type ApiType = api::RepositoryCommitsResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryCommitsResource {
            commits: self.commits.into_api(),
        }
    }
}

impl IntoApi for CommitResponse {
    type ApiType = api::RepositoryCommitResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryCommitResource {
            owner_name: self.owner_name,
            repo_name: self.repo_name,
            sha: self.sha,
            parent_sha: self.parent_sha,
            message: self.message,
            date: self.created_at,
            author: api::CommitAuthorResource {
                id: self.author_id,
                name: self.author_name,
                git_name: self.git_author_name,
                email: self.git_author_email,
                image_updated_at: self.author_image_updated_at,
            },
            review_number: self.review_number,
            diff_position: self.diff_position,
            diffs: self.diffs.into_iter().map(|d| d.into_api()).collect(),
        }
    }
}

impl IntoApi for RepositoryCommitResponse {
    type ApiType = api::RepositoryCommitResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryCommitResource {
            owner_name: String::new(),
            repo_name: String::new(),
            sha: self.sha,
            parent_sha: self
                .parent_sha
                .unwrap_or_else(|| "0000000000000000000000000000000000000000".to_string()),
            message: self.message,
            date: self.date,
            author: self.author.into_api(),
            review_number: None,
            diff_position: None,
            diffs: vec![],
        }
    }
}

impl IntoApi for CommitAuthorResponse {
    type ApiType = api::CommitAuthorResource;
    fn into_api(self) -> Self::ApiType {
        api::CommitAuthorResource {
            id: self.id,
            name: None,
            git_name: self.name,
            email: self.email,
            image_updated_at: None,
        }
    }
}

impl IntoApi for RepositoryFileResponse {
    type ApiType = api::RepositoryFileResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryFileResource {
            commit_sha: self.commit_sha,
            path: self.path,
            sha: self.sha,
            content: self.content,
            encoding: self.encoding,
        }
    }
}

impl IntoApi for RepositoryFolderResponse {
    type ApiType = api::RepositoryFolderResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryFolderResource {
            commit_sha: self.commit_sha,
            path: self.path,
            entries: self.entries.into_api(),
        }
    }
}

impl IntoApi for RepositoryBlobsResponse {
    type ApiType = api::RepositoryBlobsResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryBlobsResource {
            blobs: self.blobs.into_api(),
        }
    }
}

impl IntoApi for RepositoryBlobResponse {
    type ApiType = api::RepositoryBlobResource;
    fn into_api(self) -> Self::ApiType {
        match self {
            RepositoryBlobResponse::File(f) => api::RepositoryBlobResource::File(f.into_api()),
            RepositoryBlobResponse::Folder(f) => api::RepositoryBlobResource::Folder(f.into_api()),
        }
    }
}

impl IntoApi for RepositoryPathsResponse {
    type ApiType = api::RepositoryPathsResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryPathsResource {
            ref_name: self.ref_name,
            commit_sha: self.commit_sha,
            entries: self.entries.into_api(),
        }
    }
}

impl IntoApi for RepositoryPath {
    type ApiType = api::RepositoryPathResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryPathResource {
            path: self.path,
            name: self.name,
            path_type: self.path_type.into_api(),
            sha: self.sha,
        }
    }
}

impl IntoApi for PathType {
    type ApiType = api::PathType;
    fn into_api(self) -> Self::ApiType {
        match self {
            PathType::Blob => api::PathType::Blob,
            PathType::Tree => api::PathType::Tree,
            PathType::Commit => api::PathType::Commit,
            PathType::Unknown => api::PathType::Unknown,
        }
    }
}

impl IntoApi for CommitDiff {
    type ApiType = api::RepositoryDiffStatResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryDiffStatResource {
            path: self.path,
            lines_added: self.lines_added as u32,
            lines_removed: self.lines_removed as u32,
        }
    }
}

impl IntoApi for RepositoryBlobPairResponse {
    type ApiType = api::RepositoryBlobPairResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryBlobPairResource {
            path: self.path,
            old: self.old.map(IntoApi::into_api),
            new: self.new.map(IntoApi::into_api),
        }
    }
}

impl IntoApi for RepositoryBlobDiffsResponse {
    type ApiType = api::RepositoryBlobDiffsResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryBlobDiffsResource {
            diffs: self
                .diffs
                .into_iter()
                .map(|(k, v)| (k, v.into_api()))
                .collect(),
        }
    }
}

impl IntoApi for RepositoryDiffFileResponse {
    type ApiType = api::RepositoryDiffFileResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryDiffFileResource {
            path: self.path,
            lines_added: self.lines_added,
            lines_removed: self.lines_removed,
            left_content: self.left_content,
            right_content: self.right_content,
        }
    }
}

impl IntoApi for RepositoryActivityEvent {
    type ApiType = api::RepositoryActivityEventResource;
    fn into_api(self) -> Self::ApiType {
        match self {
            RepositoryActivityEvent::Starred { user, at } => {
                api::RepositoryActivityEventResource::Starred {
                    user: user.into_api(),
                    at,
                }
            }
        }
    }
}

impl IntoApi for RepositoryCommitFilterResponse {
    type ApiType = api::RepositoryCommitFilterResource;
    fn into_api(self) -> Self::ApiType {
        api::RepositoryCommitFilterResource {
            id: self.id,
            repository_id: self.repository_id,
            name: self.name,
            authors: self.authors,
            tags: self.tags,
            paths: self.paths,
            created_at: self.created_at,
            updated_at: self.updated_at,
        }
    }
}
